### ISSUE TITLE
Feature Request: survey pattern

### ISSUE THREAD

**Issue by mdevan** [nats-io/gnatsd#595]
#### Use Case:

Consider an agent-like daemon, which we'll call a "minion", running on each node in a deployment. Each minion subscribes to the topic "minion.{hostname}". I can send a request (like "get uptime") to a specific minion and it replies ("2 days 23 hours"). So far, so good.

Now I'd like to talk to **all** the running minions. Currently, there is only one way to do this within NATS:

1. Have each minion subscribe to both "minion.ALL" and "minion.{hostname}". The request for which all minions are expected to respond will be sent on the "minion.ALL" topic. (Alternatively, minion can listen on "minion.*" and respond if the 2nd level matches it's hostname, which is effectively the same thing.)
2. The publisher will create a temporary inbox, and publish a request to "minion.ALL". **He then has to wait for some time**, and collect all the responses received into this inbox.

The publisher **necessarily** has to wait for a fixed time, because he does not how many topics there are, and therefore does not now when to stop receiving. If for example, the publisher knows that there are 3 minions, then he can unsubscribe after receiving 3 messages.

TL;DR: Currently the publisher does not know the subscriber count of the topic he's publishing to. Knowing this will facilitate efficient 1-to-many communication.
 
#### Proposed Change:

Have an API (let's call it Survey), which functions like this at the client-side:

```go
func (nc *Conn) Survey(subj string, data []byte, timeout time.Duration) ([]*Msg, error) {
/*
    create temporary inbox
    ChanSubscribe(tempinbox, ch)
    PublishRequest(subj, tempinbox, data)
    go func() {
        for { select {
            // timeout reached -> return
            // valid reply received -> append to the []*Msg
            // "end-of-subscribers" reply received -> return // **MAGIC!**
        } }
    } ()
    wait for goroutine
*/
}
```

The line marked MAGIC has to be supported from the server-side -- the server should somehow detect that it has received at least one reply for all subscribers to the topic, and send a "end-of-subscribers" message.

Note that this more a description of the desired end result rather than a design proposal. For example, the following might also be feasible:

```go
func (nc *Conn) GetSubscriberCount(subj string) (count int, err error) {}

// the client will then subn.AutoUnsubscribe(count)
```

(Although the subscriber count might change between the GetSubscriberCount and AutoUnsubscribe.)

#### Who Benefits From The Change(s)?

Um, anyone with such use cases as above?

Currently, NATS supports 1-to-1 ("request-reply") and 1-to-any ("load-balance") patterns efficiently. Anyone who needs to do 1-to-many ("survey") pattern would benefit.

#### Alternative Approaches

The only alternative is for the minions to register their unique topic names in etcd/consul/zookeeper. The publisher will then fetch the list of topic names from there, and then publish separate requests to each topic.

**Comment by ColinSullivan1** [comment]
@mdevan , Thank you for raising this issue.  This is a common pattern we often encounter questions about, and what this really boils down to is knowing about responders a priori to making your request. 

There is another alternative - your publishing application can maintain a list who is available to respond to its requests. This can be achieved by having subscribing applications send heartbeats advertising their presence and providing metadata (their subjects, state, etc), or periodically polling out of band and gathering responses to keep your list up to date. A benefit here is that this information often provides value beyond this particular use case.

Naturally, an application may disappear and your local list of available minions would become inconsistent with your system state (resulting in hitting a timeout in a request).  This can happen regardless of whether you are maintaining a list of potential responders in your application or the server provides it, so we prefer to let applications handle this.

**Comment by derekcollison** [comment]
I have one pattern that I have used that avoids the accounting of subscribers on the publisher side. It is designed to NOT be perfect every time.

Break the request circuit in the publisher when the delta between last received and current time reaches a threshold. Usually all of the responses will all come back in a very short time window. You can track this and adapt, but essentially you run a timer that if it fires and no messages have been received in that window, assume you are done.
